### PR TITLE
driver xt_client test: extra option to run with the adapter simulator

### PR DESCRIPTION
**Pull request by pieleric**
Use the same was as the orsay test case: add an extra state to
TEST_NOHW=sim. This indicate we are testing agains the simulator.

In such case, all test cases are safe to run, so run them all.
Also fix a few test cases to be compatible with the simulator.

### CODE REVIEW

File: src/odemis/driver/test/xt_client_test.py [hunk 1]
```diff
@@ -609,7 +616,7 @@ def test_stigmator(self):
         self.assertEqual(self.microscope.get_stigmator()[0], stig_x)
         self.assertEqual(self.microscope.get_stigmator()[1], stig_y)
         # Try to set value outside of range
-        with self.assertRaises(OSError):
+        with self.assertRaises((OSError, ValueError)):
```
Comment (tepals) [inline src/odemis/driver/test/xt_client_test.py:619]: I thought Pyro5 did not transfer ValueErrors, are you sure you need this?

I'm pretty sure the adapter only returns OSErrors. So if this is for the simulator, maybe we should update the simulator to always return OSErrors.

File: src/odemis/driver/test/xt_client_test.py [hunk 2]
```diff
@@ -31,13 +31,19 @@
...
 logging.basicConfig(level=logging.INFO)
 
-TEST_NOHW = (os.environ.get("TEST_NOHW", 0) != 0)
+# Accept three values for TEST_NOHW
+# * TEST_NOHW = 1: not connected to anything => skip most of the tests
+# * TEST_NOHW = sim: xtadapter/server_sim.py running on localhost
+# * TEST_NOHW = 0 (or anything else): connected to the real hardware
+TEST_NOHW = os.environ.get("TEST_NOHW", 0)  # Default to Hw testing
+if TEST_NOHW != "sim":
+    TEST_NOHW = (TEST_NOHW == "1")  # convert values other than "sim" to False/True
```
Comment (tepals) [inline src/odemis/driver/test/xt_client_test.py:46]: This feels like your hiding things. If someone accidentally has typed "sym", `TEST_NOHW` will be set to False, which implies hardware testing.

File: src/odemis/driver/test/xt_client_test.py [hunk 3]
```diff
@@ -381,11 +389,11 @@ def test_set_scan_field_size(self):
         init_scan_size = self.microscope.get_scanning_size()[0]
         new_scanfield_x = scanfield_range['x'][1]
         self.microscope.set_scanning_size(new_scanfield_x)
-        self.assertEqual(self.microscope.get_scanning_size()[0], new_scanfield_x)
+        self.assertAlmostEqual(self.microscope.get_scanning_size()[0], new_scanfield_x)
         # Test it still works for different values.
         new_scanfield_x = scanfield_range['x'][0]
         self.microscope.set_scanning_size(new_scanfield_x)
-        self.assertEqual(self.microscope.get_scanning_size()[0], new_scanfield_x)
+        self.assertAlmostEqual(self.microscope.get_scanning_size()[0], new_scanfield_x)
```
Comment (tepals) [inline src/odemis/driver/test/xt_client_test.py:396]: By default assertAlmostEqual rounds to 7 decimal places. The minimum range for the scanfield is 1e-8, so I think you want to test at least for 8 decimal places.

https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertAlmostEqual

File: src/odemis/driver/test/xt_client_test.py (hunk 1 above)
Comment (pieleric) [reply src/odemis/driver/test/xt_client_test.py:619]: Yes, that's for the simulator. I noticed that the current normal code either raise ValueError when the Python code checks for the bounds or OSError when it's XTLib/XTOM which checks. I concluded that both should be supported in the client, as it's hard to guess. So I've added it here, and decided to use ValueError in the simulator, as we check in the Python code, and didn't feel it's better to simulate a OSError. I can switch the simulator to OSError and here to just OSError if you see a good reason for it.

File: src/odemis/driver/test/xt_client_test.py (hunk 2 above)
Comment (pieleric) [reply src/odemis/driver/test/xt_client_test.py:46]: So, you mean it'd be better to say "anything different from 0" is NOHW testing? Or it should raise an error when it's neither sim 0 or 1? So far, in the other test cases, we consider anything different from 0 is NOHW testing.

File: src/odemis/driver/test/xt_client_test.py (hunk 3 above)
Comment (pieleric) [reply src/odemis/driver/test/xt_client_test.py:396]: Good catch. Indeed, with microscopes, small is sometimes still very big!
Our own method almost_equal uses 1e-18, which is a little safer.

File: src/odemis/driver/test/xt_client_test.py (hunk 1 above)
Comment (tepals) [reply src/odemis/driver/test/xt_client_test.py:619]: Let's leave the ValueError then.

File: src/odemis/driver/test/xt_client_test.py (hunk 2 above)
Comment (tepals) [reply src/odemis/driver/test/xt_client_test.py:46]: Maybe the double negations are messing with my mind, but the way I read this is that anything different from 0 is HW testing.
```
TEST_NOHW = "something-different"
TEST_NOHW = (TEST_NOHW == "1")  = ("something-different" == "1") = False
```

TEST_NOHW = False => TEST_HW = True

Anyway, I would say it should raise an error for anything different than sim, 0 or 1.

Comment (pieleric) [reply src/odemis/driver/test/xt_client_test.py:46]: Yes. Right now it says "anything different from 1 is normal hardware testing". That's opposite to every other test cases. I'll change it to raise an error if it's different from the 3 values.

**Review by tepals — APPROVED** [review]

**Pull request merged**